Add character-level diff for strings.

roberthoenig opened the pull request:
Implements #91.

@josephburnett this is my first attempt at getting character-level diff to work, and by no means a complete PR. 

Can you take a look and see if this goes in the right direction?

What this PR does:
``` diff
+ Implement a character-level diff for strings
```

What this PR does not do (yet):
``` diff
- nicely structure code
- add tests
- add a command-line option
- implement a more optimal diff algorithm (Myers diff for maximizing the common sequence overlap.)
```

## Example:

`a.json`:
   ```
   {
    "foo": "bar",
    "baz": 1,
    "qux": ["foobar", "foobaz"]
   }
   ```
`b.json`:
```
{
    "foo": "bazzz",
    "baz": 2,
    "qux": ["foarrr", "foobaz"]
}
```

`jd -v2 a.json b.json`:
<img src="https://github.com/user-attachments/assets/7ba90c01-12fa-4e36-8f10-91168df412a8" width="96">

josephburnett commented:
@roberthoenig this is looking very good!

You probably already encountered this, but one awkward thing in testing this code is that `main.go` is in the v1 module while new changes (like yours) are in the `v2` module. So in order to pick them up locally a `replace` directive must be added temporarily to the `go.mod` file:

```
replace github.com/josephburnett/jd/v2 => ./v2
```

I'll be moving `main.go` into `v2` eventually which will remove this annoyance and should also fix https://github.com/josephburnett/jd/issues/90.

Thanks for the nice description and list if what's not there. Yeah, the code could use a few descriptive functions to reduce the nesting. And tests, yes. I'm not too worried about the sub-optimal nature of the diff, more about bugs. So if you want to implement a more efficient algorithm yourself we should have a very high bar for unit tests. But also I would want to look at the performance of the algorithms over the wide range of use cases jd is applied to. My preference is to leave it as-is, but I'm not saying "no" :wink: 

I wonder if we should just leave out the option and default to character-level color diffs. The color is a presentation level feature and I would not expect anyone to be relying on the color codes (and I would feel comfortable saying "don't, they might change"). If they DO want character-level diff data, we should be descending into the string during diff (which is totally possible) and producing character-level hunks.

josephburnett commented:
Also you don't need to set `-v2` because it's on by default now.

roberthoenig commented:
Thanks for the quick feedback! I'll try to get a mergeable version ready, then. Got quite a few things on my plate right now though, so might take until next week or so.

roberthoenig commented:
@josephburnett I've updated the code into a (hopefully) mergeable version.

* I've not added a separate option to enable character-level diffs. Instead, they're automatically enabled when using `-color`.

* I could not yet figure out how to run the tests for v2. I'm using
   ```
   replace github.com/josephburnett/jd/v2 => ./v2
   ```
   in the main go.mod (since it's also necessary to build locally). But this seems to have no effect on the tests for v2? They always pass locally, even though CI fails. I've also tried adding `replace github.com/josephburnett/jd/v2 => ../v2` to `v2/go.mod`, to no avail.

ccoVeille commented on v2/diff_write.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/json"
 	"fmt"
 
+	lcs "github.com/yudai/golcs"
```
https://github.com/yudai/golcs seems deprecated, or at least unmaintained

Did you consider other libraries such as these ones?

https://github.com/vmarkovtsev/go-lcss
https://github.com/peter-evans/patience

These are examples of library I found by taking a look at available libraries

https://github.com/search?q=language%3Ago+lcs&type=repositories&s=updated&o=desc

roberthoenig replied:
github.com/yudai/golcs is the lcs library used in other places in the codebase of `jd` (see `v2/list.go`)

ccoVeille replied:
Indeed, otherwise the go.mod would have been updated 😅

So, it's of course normal to use it.

josephburnett commented:
@roberthoenig The `replace` directive is helpful for development while the `main.go` files is in the root (v1) because the root `go.mod` has a specific version of `v2` it uses (it's a separate module after all). So to pick up changes naturally in v1, you would have to commit changes to v2, push and then run `go mod tidy` or something to upgrade. The replace directive let's you test with v1 main without any of that nonsense. This problem will go away (soon) when I move main to v2.

The unit tests in each module can be run separately by running `go test` in each. Easiest is just to run `make test` which will change directories for you. You don't need the replace directive to make the unit tests pass.

@ccoVeille The LCSS is a different algorithm I think. It does something different. The Patience algorithm I'm not familiar with. I would use the same considerations as I mentioned above for Robert. I would want to compare the efficiency of the algorithm across the wide range of use cases jd is applied to. But for now, I don't really have a problem with the current library. It's not very complex and not any significant attack surface, so I don't much reason to change. Do you have another reason I haven't though of? Thanks for checking that out!

josephburnett commented on v2/diff_write.go:
```diff
@@ -22,6 +42,28 @@ func (d DiffElement) Render(opts ...Option) string {
 	b.Write([]byte(d.Path.JsonNode().Json()))
 	b.WriteString("\n")
+
+	// Check if this is a single string diff. If so, compute the common seqeuence for a character
```
Nit: s/seqeuence/sequence/

roberthoenig commented on v2/diff_write.go:
```diff
@@ -36,40 +78,63 @@ func (d DiffElement) Render(opts ...Option) string {
-		if !isVoid(oldValue) {
+		if isSingleStringDiff && isColor {
+			oldStr := string(oldValue.(jsonString))
+			b.WriteString(" asdfasdfasd\"")
```
:grimacing: my debug string made it into the PR -- luckily, my newly added tests discovered this bug!

roberthoenig commented:
@josephburnett updated the PR:

* fixed the typo
* added tests for colored diffs
   * first, the new test renders the diff with `color` enabled, and verifies that the resulting diff is correct, when the color is ignored.
   * then, the new test verifies that the uncolored parts of string diffs match.

Are there any other changes you wanna see, or is this good to go now?

josephburnett commented:
> Are there any other changes you wanna see, or is this good to go now?

This looks great @roberthoenig. Thank you for the contribution!

roberthoenig commented:
Awesome! @josephburnett would it be possible to do a new release with this feature now, so that we can use it?

josephburnett commented:
@roberthoenig okay! I've release `2.2.1` which has your color changes. Also forked `main.go` so the latest v2 can be picked up with `go install github.com/josephburnett/jd/v2/jd@latest`.
